salesforce/bulkwrite.go

```diff
@@ -246,35 +271,124 @@ func (c *Connector) GetJobInfo(ctx context.Context, jobId string) (*GetJobInfoRe
 	return info, nil
 }
 
-/*
-// Below are methods that will help us get results, but need CSVHTTPClient to be implemented
-// Below code will work once CSVHTTPClient is implemented
-// TODO: implement CSVHTTPClient
-
-func (c *Connector) GetJobResult(ctx context.Context, jobId string) ([]byte, error) {
-	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/successfulResults", jobId))
+func (c *Connector) GetJobResults(ctx context.Context, jobId string) (*JobResults, error) {
+	jobInfo, err := c.GetJobInfo(ctx, jobId)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get job information: %w", err)
 	}
 
-	return c.getCSV(ctx, location)
+	if jobInfo.State != JobStateComplete {
+		// Take care of failed, aborted, in progress, and upload complete cases
+		// We don't need to query SF for these cases
+		return c.getUnsuccessfulJobResults(jobInfo), err
+	}
+
+	if jobInfo.State == JobStateComplete && jobInfo.NumberRecordsFailed == 0 {
+		// Complete success case, no need to query SF
+		return &JobResults{
+			JobId:   jobInfo.Id,
+			State:   jobInfo.State,
+			JobInfo: jobInfo,
+		}, nil
+	}
+
+	return c.getPartialFailureDetails(ctx, jobInfo)
 }
 
-func (c *Connector) GetUnprocessedResults(ctx context.Context, jobId string) ([]byte, error) {
-	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/unprocessedrecords", jobId))
+func (c *Connector) getJobResults(ctx context.Context, jobId string) (*http.Response, error) {
+	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/failedResults", jobId))
 	if err != nil {
 		return nil, err
 	}
 
-	return c.getCSV(ctx, location)
+	req, err := common.MakeJSONGetRequest(ctx, location, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create get request: %w", err)
+	}
+
+	return c.Client.Client.Do(req)
 }
 
-func (c *Connector) GetFailedResults(ctx context.Context, jobId string) ([]byte, error) {
-	location, err := joinURLPath(c.BaseURL, fmt.Sprintf("jobs/ingest/%s/failedResults", jobId))
+func (c *Connector) getPartialFailureDetails(ctx context.Context, jobInfo *GetJobInfoResult) (*JobResults, error) {
+	// Query SF to get partial failure details
+	res, err := c.getJobResults(ctx, jobInfo.Id)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get job results: %w", err)
+	}
+	defer res.Body.Close()
+	// if partially failed, we need to get the error message for each record
+	reader := csv.NewReader(res.Body)
+
+	failInfo := &FailInfo{
+		FailureType:   "Partial",
+		FailedRecords: make(map[string][]string),
+	}
+
+	sfIdIdx := -1
+	sfErrorIdx := -1
+	idx := 0
+
+	for {
+		record, err := reader.Read()
+		if err != nil {
+			// end row
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return nil, err
+		}
+
+		if idx == 0 {
+			for idx, column := range record {
+				switch column {
+				case "sf__Id":
+					sfIdIdx = idx
+				case "sf__Error":
+					sfErrorIdx = idx
+				}
+			}
+
+			idx++
+
+			continue
+		}
```

Comment:
What happens if `sf__Id` or `sf__Error` is missing?

Reply:
this still looks unresolved, now `sfIdIdx` and `sfErrorIdx` default to 0, there needs to be a check still